### PR TITLE
Introduction and overview 

### PR DESCRIPTION
**Pull request by abitha-thankaraj**
TODO 

- [x] Fix formatting
- [x] Add references
- [x] Squash commits

### CODE REVIEW

File: feb09/at4524.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+- *Data shuffling*. Careful attention must be paid to the implementation of the data-shuffling phase to avoid generating O(M\*R) seeks in a MapReduce with M map tasks and R reduce tasks;
+- *Textual formats*. MapReduce users should avoid using inefficient textual formats;
+- *Natural indices*. MapReduce users should take advantage of natural indices (such as timestamps in log file names) whenever possible; and
+- *Unmerged output*. Most MapReduce output should be left unmerged, since there is no benefit to merging if the next consumer is another MapReduce program.
+
+
+[[https://www.cs.cmu.edu/\~pavlo/slides/mapreduce-pavlo09.pdf]{.ul}][14]
+[[https://cacm.acm.org/magazines/2010/1/55744-mapreduce-a-flexible-data-processing-tool/fulltext]{.ul}][5]
+MapReduce in Practice •Abstract input and output interfaces --lots of MR operations don't just read/write simple files • B-tree files • memory-mapped key-value stores • complex inverted index file formats • BigTable tables • SQL databases, etc. • \... •Low-level MR interfaces are in terms of byte arrays --Hardly ever use textual formats, though: slow, hard to parse --Most input & output is in encoded Protocol Buffer format\
```
Comment (jvenkit1) [inline feb09/at4524.md]: ~~nitpicking here, but the formatting seems to be off for this line.~~ nvm, just read the TODO.

**Comment by jvenkit1** [comment]
Great notes !

I am not able to preview the images. Dunno why, but it doesn't load and the links give a 404 error.

File: feb09/at4524.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+[The Datacenter as a Computer (3rd Ed.)](https://cs.nyu.edu/~apanda/classes/sp21/papers/dc-computer-v3.pdf) Ch 1 provides an overview of the primary components of a WSC namely -  server, storage, networking and the challenges that accompany its massive scale. Ch 2. Explores common software methods used to ensure high performance in large distributed systems. 
+
+![Evolution of large scale computing](images/evolution-of-large-scale-computing.jpeg)
+
+Source: a16z
+
+With the influx of large amounts of data on the web, due to widespread availability of internet, Google developed a scalable distributed file system for large distributed data intensive applications - namely, [GFS.](https://cs.nyu.edu/~apanda/classes/sp21/papers/gfs.pdf) [MapReduce](https://cs.nyu.edu/~apanda/classes/sp21/papers/mapreduce.pdf) was introduced to parallelize simple computational tasks for processing and generating large
+data sets. while abstracting out the messy underlying details of parallelization, fault-tolerance, data distribution
+and load balancing in a library.
+<!------
```
Comment (apanda) [inline feb09/at4524.md]: Did you mean to comment out what follows? If so can you delete the comment before merging.

File: feb09/at4524.md (hunk 1 above)
Comment (apanda) [reply feb09/at4524.md]: Are you planning on actually addressing the TODOs, they seem to be pending for several days.

File: feb09/at4524.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+  [6]: media/image3.png {width="2.8958333333333335in" height="2.929134951881015in"}
+  [7]: media/image2.png {width="3.03125in" height="2.9166666666666665in"}
+  [8]: https://pierrezemb.fr/posts/colossus-google/
+  [9]: https://cloud.google.com/files/storage_architecture_and_challenges.pdf
+  [10]: https://en.wikipedia.org/wiki/Reed%E2%80%93Solomon_error_correction
+  [11]: https://queue.acm.org/detail.cfm?id=1594206
+  [12]: http://www.pdsw.org/pdsw-discs17/slides/PDSW-DISCS-Google-Keynote.pdf
+  [13]: https://medium.com/@oscarstiffelman/a-brief-history-of-mapreduce-97aec97df8ff
+  [14]: https://www.cs.cmu.edu/~pavlo/slides/mapreduce-pavlo09.pdf
+----->
```
Comment (apanda) [inline feb09/at4524.md]: Why are cites commented? It is generally good form to explicitly include the cites in text.

File: feb09/at4524.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    -   GFS master replaced by Colossus
+    -   GFS chunkserver replaced by D
+    -   Colossus rebalances old, cold data
+    -   distributes newly written data evenly across disks
+    -   Metadata is stored into BigTable. each Bigtable row corresponds to a single file. Adoption of BigTable provides several advantages
+        -   It removed the single point of failure in GFS master
+        -   Automatic sharding of regions
+        -   Lexical ordering
+
+With the Google Home Ecosystem being widely adopted, New offerings like [Anthos](https://cloud.google.com/solutions/anthos-edge) (hybrid cloud management) and [Edge TPU](https://cloud.google.com/edge-tpu) look promising for edge computing. Google's file storage will now have to support latency sensitive applications.
```
Comment (apanda) [inline feb09/at4524.md]: But is GFS the only file system Google uses? What would you use for these use cases. How would trade-offs change?

**Review by apanda — COMMENTED** [review]
Overall, I think what is **good** about this write up is that you really found sources outside; my big **concern** is that your analysis of these outside sources is very shallow. For example you talk about the edge, cool, tell us what changes in your opinion, or how you would change things.

I am also a little concerned about formatting, do you know when you will get to the TODOs.

File: feb09/at4524.md (hunk 3 above)
Comment (abitha-thankaraj) [reply feb09/at4524.md]: The formatting was off. WIll resolve this.

File: feb09/at4524.md (hunk 2 above)
Comment (abitha-thankaraj) [reply feb09/at4524.md]: WIll delete this before merge.

File: feb09/at4524.md (hunk 1 above)
Comment (abitha-thankaraj) [reply feb09/at4524.md]: Apologies. WIll commit them by tonight.

File: feb09/at4524.md (hunk 4 above)
Comment (abitha-thankaraj) [reply feb09/at4524.md]: Added details in next commit.

File: feb09/at4524.md [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    -   Recovery - Shadow master - read-only access to the file system, even when primary master is down. They read a replica of the operation log and apply the same sequence of changes. Master recovers from checkpoints on operation log.
+-   Chunkserver
+    -   Store the chunks (64 MB)on local disks as Linux files
+    -   Metadata per chunk is \< 64 bytes (stored in master)
+        -   Current replica locations
+        -   Reference count (useful for copy-on-write)
+        -   Version number (for detecting stale replicas)
+    -   Fully replicated (for high availability) - default replication factor:3
+-   Handling writes
+    -   Record append allows multiple clients to append data to the same file concurrently while guaranteeing the atomicity of each individual client's append.
```
Comment (neeraj71) [inline feb09/at4524.md:116]: How's the record append done?

Comment (abitha-thankaraj) [reply feb09/at4524.md:116]: Lease mechanism coordinated by the master node. Didnt discuss this in detail because I wasn't able to outline all the approaches succinctly
